### PR TITLE
Convert the 1-way SSE4 SHA256 code from asm to intrinsics

### PR DESCRIPTION
**Pull request by sipa**
Currently, master contains 2 implementations of SHA256 for SSE4:
* A generic one written using GCC inline assembly (converted from Intel NASM code), added in #10821.
* A specialized double-SHA256 for 64-byte inputs written using intrinsics, added in #13191.

The advantage of the inline assembly is that its performance is not affected by compiler optimizations (and doesn't even need compiler support for SSE4). The downside is that it is an opaque, unreadable, non-reusable blob of code.

This patch converts the former also to intrinsics - making its operation more clear, while hopefully lending itself to being adaptable for other specialized implementations.

The resulting implementation is slightly faster on my system (i7-7820HQ) when compiled with GCC 7.3. Small variations in the code can affect the optimizer though, and have as much as a few % impact on speed.

### CODE REVIEW

**Comment by theuni** [comment]
Nice!

@sipa See https://github.com/theuni/bitcoin/commit/d69a5164f914c6c2945d8f32134faa0da87795f5 for clang compile fixes, and https://github.com/theuni/bitcoin/commit/4ed6f4fc901f8f83e3cc4c48ebb678ff795c03c1 for a change that may or may not be needed to avoid a performance hit on AMD.

File: src/crypto/sha256_sse41.cpp [hunk 1]
```diff
@@ -10,6 +14,221 @@
...
+    const __m128i SHUF_00BA = _mm_set_epi64x(0xFFFFFFFFFFFFFFFFULL, 0x0b0a090803020100ULL);
+    const __m128i SHUF_DC00 = _mm_set_epi64x(0x0b0a090803020100ULL, 0xFFFFFFFFFFFFFFFFULL);
+
+    W = _mm_add_epi32(W, X0);
+    _mm_store_si128((__m128i*)Ws, W);
+
+    Round(a, b, c, d, e, f, g, h, Ws[0]);
+    XTMP0 = _mm_alignr_epi8(X3, X2, 4);
+    XTMP0 = _mm_add_epi32(XTMP0, X0);
+    XTMP3 = XTMP2 = XTMP1 = _mm_alignr_epi8(X1, X0, 4);
```
Comment (theuni) [inline src/crypto/sha256_sse41.cpp]: Is there some voodoo here, why not just use XTMP3 below? Does this avoid a pipeline stall or something?

**Comment by sipa** [comment]
@theuni Included the clang compile fixes. I'm going to benchmark to see whether to include the other changes.

File: src/crypto/sha256_sse41.cpp (hunk 1 above)
Comment (sipa) [reply src/crypto/sha256_sse41.cpp]: No idea. It's just a translation of the existing assembly code.

**Comment by sipa** [comment]
It would be worthwhile to benchmark this on reasonably recent clang versions as well - the performance impact may be very different depending on how good the compiler is at ordering parallel instruction paths.

**Comment by sipa** [comment]
Some more benchmarks, comparing GCC 7.3 and clang 6.0, for the `SHA256` benchmark:

* GCC, master: 4.4 ms
* GCC, this PR: 4.3 ms
* clang, master: 4.4 ms
* clang, this PR: 4.8 ms

Unfortunately, it seems that clang isn't as good in producing as performant code from intrinsics.

**Comment by theuni** [comment]
@sipa Mind rebasing? I'd like to add the lib-per-cpu changes on top of this.

**Comment by sipa** [comment]
Rebased, though I don't think this PR is acceptable until we have a way to avoid the performance loss in clang.

File: src/crypto/sha256_sse41.cpp [hunk 2]
```diff
@@ -6,6 +10,211 @@
...
+ */
+void inline __attribute__((always_inline)) QuadRoundSched(uint32_t& a, uint32_t& b, uint32_t& c, uint32_t& d, uint32_t& e, uint32_t& f, uint32_t& g, uint32_t& h, __m128i& x0, __m128i x1, __m128i x2, __m128i x3, __m128i w)
+{
+    alignas(__m128i) uint32_t w32[4];
+    __m128i t0, t1, t2, t3, t4;
+
+    w = _mm_add_epi32(w, x0);
+    _mm_store_si128((__m128i*)w32, w);
+
+    Round(a, b, c, d, e, f, g, h, w32[0]);
```
Comment (practicalswift) [inline src/crypto/sha256_sse41.cpp]: is value read from `w32` is guaranteed to be initialised here?

File: src/crypto/sha256_sse41.cpp [hunk 3]
```diff
@@ -6,6 +10,211 @@
...
+    t0 = _mm_add_epi32(_mm_alignr_epi8(x3, x2, 4), x0);
+    t3 = t2 = t1 = _mm_alignr_epi8(x1, x0, 4);
+    t2 = _mm_srli_epi32(t2, 7);
+    t1 = _mm_or_si128(_mm_slli_epi32(t1, 32 - 7), t2);
+
+    Round(h, a, b, c, d, e, f, g, w32[1]);
```
Comment (practicalswift) [inline src/crypto/sha256_sse41.cpp]: Same here and throughout this function :-)

File: src/crypto/sha256.cpp [hunk 4]
```diff
@@ -615,12 +615,9 @@ std::string SHA256AutoDetect()
 #endif
 
     if (have_sse4) {
```
Comment (practicalswift) [inline src/crypto/sha256.cpp]: Move the `if` statement inside of the `#if defined(ENABLE_SSE41) && !defined(BUILD_BITCOIN_INTERNAL)` to remove the possibility of an empty if statement.

**Comment by STALININST** [comment]
Please leave me out of this coversation, i have no idea what are you 
doing and please give me access to bitcoin core repository, i will check 
if there is something i can help with

**Comment by MarcoFalke** [comment]
Are you still working on this?

**Comment by sipa** [comment]
What version of clang are we using now? It's probably not a good idea to proceed with this unless it can be shown it doesn't have negative impact on performance on all release platforms.

**Comment by MarcoFalke** [comment]
* 3.7 (#16052) for gitian
* FreeBSD 12: clang version 6.0.1
* not sure what the default is on macos when self-compiled

File: src/crypto/sha256_sse41.cpp (hunk 2 above)
Comment (sipa) [reply src/crypto/sha256_sse41.cpp]: Yes, in the line above.

**Comment by sipa** [comment]
I'll close this for now, then.

**Comment by fanquake** [comment]
@MarcoFalke, Clang on my macOS machine (Xcode 10.2.1) is:
```
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by DesWurstes** [comment]
...which is actually Clang 5.

**Comment by fanquake** [comment]
I'm going to reopen this, as we will be switching to a newer version of `Clang` in gitian.

**Comment by sipa** [comment]
I'll benchmark again in clang-7.

**Comment by MarcoFalke** [comment]
I see a slowdown in `SHA256` and `SHA256_32b` with both gcc-9 and clang-8

**Comment by fanquake** [comment]
Futher benchmarking reported here and outside this PR have shown that there are likely slowdown issues with this change and recent versions of Clang. Closing again for now.

**Comment by hebasto** [comment]
Picked up for MSVC builds in https://github.com/bitcoin/bitcoin/pull/28526.

**Pull request closed without merging**